joomla/joomla-cms
# Session interaction refactoring

#### Summary of Changes

To try and sum things up:
- `JSession` can now lazy start a session; calls to `get()`, `set()`, `has()`, and `clear()` will start the session if it hasn't explicitly been started previously
- The application classes are restructured to not explicitly start the session; this defers the point `JSession::start()` is called from the application constructors to one of the first actions done after `$app->execute()` is called.  The advantage here is that now sessions are started after the application has been instantiated and set to `JFactory::$application`.
- The logic for purging of session metadata is altered so that this step only occurs if the database session handler is not in use and it is now executed as part of the `onAfterRender` event instead of when the session is originally started; garbage collection will occur for the database session handler as part of the PHP session API
- `JApplicationCms::checkSession()` now optionally accepts the `JSession` object that should be checked and this will be a mandatory argument as of 4.0.  This method is called as part of the `afterSessionStart` event which may occur without a session object stored to `JFactory`.
- The `afterSessionStart` event now includes the `JSession` object for which the session was started for as an argument of the event, this should be used instead of fetching the session from `JFactory`
- Adjustments to the database session handler:
  - `JSessionStorageDatabase` is dependent on the application to insert the session record into the database table, the logic is added to the `write()` method to remove this dependency
  - Garbage collection when the database session store is used is deferred to the end of the request cycle to help prevent a request failure solely because expired session records are being purged.  As explained in the [PHP documentation](http://php.net/manual/en/sessionhandler.gc.php) the `gc()` method is called when the session is started.
  - `JApplicationCms::checkSession()` is adjusted to only manipulate the `time` column when not using the database session handler to preserve the record's integrity when written by the session API
#### Testing Instructions

1) Download the branch from https://github.com/mbabker/joomla-cms/archive/refactor-session-metadata-coupling.zip
2) Install Joomla; sessions in the installation application should continue to function correctly
3) Use Joomla normally; sessions should continue to function correctly
4) With the database session handler enabled, `JApplicationCms::checkSession()` should only be executed when the session is new.  As this is only ensuring the extra metadata is present in the database, it is not required for every page load.  All requests after the session is started will have the record in the database updated by the session handler only (or any place making an explicit query to the `#__session` database table).  As well, the new `JApplicationCms::cleanupSessionMetadata()` method should never be called.
5) With any other session handler enabled, `JApplicationCms::checkSession()` should only be called if the session is new or the time is an even numbered second.  The new `JApplicationCms::cleanupSessionMetadata()` method should be called when the `onAfterRender` event is triggered and the delete query should only run when the time is an even numbered second.
